### PR TITLE
Choose pkn

### PR DESCRIPTION
**Pull request by pantoja**
## Description
Choose your Team to battle with your opponent. The team and the chosen pokemon will be saved in the project's database.

## Trello Card
As the opponent Trainer I select my PKNs to battle

## Screenshots (if appropriate):
image

## Steps to reproduce (if appropriate):

1. Create a superuser and login in http://127.0.0.1:8000/admin
2. Create another user to battle
3. In http://127.0.0.1:8000/create-battle/ create a new battle
4. In http://127.0.0.1:8000/create-team/ create your team typing your chosen pokemon's ids.

## Results
Check if the team, pokemon and battle are saved.

### CODE REVIEW

File: backend/battles/models.py [hunk 1]
```diff
@@ -10,6 +12,18 @@ class Battle(models.Model):
...
     def __str__(self):
         return "Battle nº " + str(self.id)
+
+
+class Team(models.Model):
+    battle = models.ForeignKey(Battle, on_delete=models.CASCADE, related_name="battle")
+    trainer = models.ForeignKey(
+        User, on_delete=models.CASCADE, verbose_name="Trainer", related_name="trainer"
+    )
+    team = models.ManyToManyField(Pokemon, verbose_name="Team")
```
Comment (pamella) [inline backend/battles/models.py]: I think you might have misunderstood the usage of the `related_name` attribute. I did too when I was at the beginning of my pokebattle.

The `related_name` attribute specifies the name of the reverse relation from the User model back to your model (line 23).

If you don't specify a `related_name`, Django automatically creates one using the name of your model with the suffix `_set`, for instance, `User.team_set.all()`.

If you do specify, e.g. `related_name=teams` on the User model, `User.team_set` will still work, but the `User.teams.` syntax is cleaner and more intuitive. So for example, if you had a user object `current_user`, you could use `current_user.teams.all()` to get all instances of your `Team` model that have a relation to `current_user`, in other words, you can get all teams from that specific user. 

**P.S.:** The logic applies not only for the `trainer` field but also for the `battle` one. Let me know if you still have doubts about the `related_name` usage, [here is a link from Django docs](https://docs.djangoproject.com/en/3.0/topics/db/queries/#following-relationships-backward) but we can chat in person as well.

---

Considering that, I would suggest the change below:

```suggestion
class Team(models.Model):
    battle = models.ForeignKey(Battle, on_delete=models.CASCADE, related_name="teams")
    trainer = models.ForeignKey(
        User, on_delete=models.CASCADE, verbose_name="Trainer", related_name="teams"
    )
    team = models.ManyToManyField(Pokemon, verbose_name="Team")
```

File: backend/battles/views.py [hunk 2]
```diff
@@ -1,16 +1,52 @@
...
     form_class = BattleForm
     model = Battle
-    success_url = reverse_lazy("home")  # TODO: Change success url to the next step
 
     def get_initial(self):
         return {"user_creator": self.request.user}
+
+    def get_success_url(self):
+        self.success_url = reverse_lazy("battles:create_team", args=[self.object.id])
+        return self.success_url
```
Comment (pamella) [inline backend/battles/views.py]: If you want you can return the URL directly without storing at some variable. It isn't wrong the way you did, I'm just pointing another option. Feel free to choose what you prefer :)

```suggestion
    def get_success_url(self):
        return reverse_lazy("battles:create_team", args=[self.object.id])
```

File: backend/services/api.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import requests
+
+
```
Comment (pamella) [inline backend/services/api.py]: If you want you can store the Poke API root URL at some global variable, so you can access easily at other scopes. Again it isn't wrong the way you did, feel free to choose what you prefer :)

```suggestion
POKEAPI_ROOT_URL = 'https://pokeapi.co/api/v2/pokemon/'
```

File: backend/services/api.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+def get_pokemon_list():
+    url = "https://pokeapi.co/api/v2/pokemon/?limit=50"
+    response = requests.get(url)
+    response = response.json()
+    return response["results"]
```
Comment (pamella) [inline backend/services/api.py]: One more suggestion and again it isn't wrong the way you did, feel free to choose what you prefer :)

```suggestion
def get_pokemon_list():
    response = requests.get(f"{POKEAPI_ROOT_URL}?limit=50")
    if not response.ok:
        return None   
    return response.json()["results"]

```

File: backend/services/api.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    response = response.json()
+    data = {
+        "name": response["forms"][0]["name"],
+        "id": response["id"],
+        "sprite": response["sprites"]["front_default"],
+        "attack": response["stats"][4]["base_stat"],
+        "defense": response["stats"][3]["base_stat"],
+        "hp": response["stats"][5]["base_stat"],
+    }
+    return data
```
Comment (pamella) [inline backend/services/api.py]: Feel free to check if the response returns status code 200 OK before accessing its JSON.

```suggestion
def get_pokemon_stats(pokemon):
    response = requests.get(f"{POKEAPI_ROOT_URL}{pokemon}")
    if not response.ok:
        return None
    response = response.json()
    data = {
        "name": response["forms"][0]["name"],
        "id": response["id"],
        "sprite": response["sprites"]["front_default"],
        "attack": response["stats"][4]["base_stat"],
        "defense": response["stats"][3]["base_stat"],
        "hp": response["stats"][5]["base_stat"],
    }
    return data
```

File: backend/battles/models.py (hunk 1 above)
Comment (pantoja) [reply backend/battles/models.py]: So this also applies to line 9 and 12 too?
`        User, on_delete=models.CASCADE, verbose_name="Creator", related_name="user_creator"
`
becomes
`        User, on_delete=models.CASCADE, verbose_name="Creator", related_name="battles"
`
and 

`        User, on_delete=models.CASCADE, verbose_name="Opponent", related_name="user_opponent"
`

becomes
`        User, on_delete=models.CASCADE, verbose_name="Opponent", related_name="battles"
`

Comment (pantoja) [reply backend/battles/models.py]: Thanks, i corrected these values and the same ones in Battle models too

Comment (pamella) [reply backend/battles/models.py]: Very nice that you added there as well! :) I've seen that you changed to `created_battle` and` invited_to_battle`, and it's kind of a pattern to use the `related_name` in the plural, but you can do that in next opportunities.

**Pull request merged**